plugins/inlinetoolbar/plugin.js

```diff
@@ -168,13 +174,412 @@
+
+		this._attachListeners();
+	}
+
+	ContextManager.prototype = {
+		/**
+		 * Adds a `context` to the tracked contexts list.
+		 *
+		 * @param {CKEDITOR.plugins.inlinetoolbar.context} context
+		 */
+		add: function( context ) {
+			this._contexts.push( context );
+		},
+
+		/**
+		 * Check each registered context against `selection` to find the best match. By default only one
+		 * toolbar per manager will be shown.
+		 *
+		 * @param {CKEDITOR.dom.selection/null} [selection=null] Selection to be used for probing toolbar. If none provided, a
+		 * _shrunk_ selection of current editor will be used.
+		 */
+		check: function( selection ) {
+			if ( !selection ) {
+				selection = this.editor.getSelection();
+
+				// Shrink the selection so that we're ensured innermost elements are available.
+				CKEDITOR.tools.array.forEach( selection.getRanges(), function( range ) {
+					range.shrink( CKEDITOR.SHRINK_ELEMENT, true );
+				} );
+			}
+
+			if ( !selection ) {
+				return;
+			}
+
+			var forEach = CKEDITOR.tools.array.forEach,
+				mainRange = selection.getRanges()[ 0 ],
+				path = mainRange && mainRange.startPath(),
+				highlightElement,
+				contextMatched;
+
+			// This function encapsulates matching algorithm.
+			function matchEachContext( contexts, matchingFunction, matchingArg1 ) {
+				forEach( contexts, function( curContext ) {
+					// Execute only if there's no picked context yet, or if probed context has a higher priority than
+					// currently matched one.
+					if ( !contextMatched || contextMatched.options.priority > curContext.options.priority ) {
+						var result = matchingFunction( curContext, matchingArg1 );
+
+						if ( !!result ) {
+							if ( result instanceof CKEDITOR.dom.element ) {
+								highlightElement = result;
+							} else {
+								// Reset what could have been previously set highlight element, as it's no longer relevant.
+								highlightElement = null;
+							}
+
+							contextMatched = curContext;
+						}
+					}
+				} );
+			}
+
+			function elementsMatcher( curContext, curElement ) {
+				return curContext._matchElement( curElement );
+			}
+
+			// Match callbacks.
+			matchEachContext( this._contexts, function( curContext ) {
+				return curContext._matchRefresh( path, selection );
+			} );
+
+			// Match widgets.
+			matchEachContext( this._contexts, function( curContext ) {
+				return curContext._matchWidget();
+			} );
+
+			// Match element selectors.
+			if ( path ) {
+				for ( var i = 0; i < path.elements.length; i++ ) {
+					var curElement = path.elements[ i ];
+					// Skip non-editable elements (e.g. widget internal structure).
+					if ( !curElement.isReadOnly() ) {
+						matchEachContext( this._contexts, elementsMatcher, curElement );
+					}
+				}
+			}
+
+			this.hide();
+
+			if ( contextMatched ) {
+				if ( !highlightElement ) {
+					highlightElement = ( path && path.lastElement ) || this.editor.editable();
+				}
+
+				contextMatched.show( highlightElement );
+			}
```

Comment:
If `matchEachContext` above can be simplified, also this part can as `contextMatched` is set only when `highlightElement ` is present so you can just go with:

```
if ( contextMatched ) {
    contextMatched.show( highlightElement );
}
```

Reply:
Now due to this part: https://github.com/ckeditor/ckeditor-dev/blob/e382734/plugins/inlinetoolbar/plugin.js#L425-L428

It's guaranteed that `highlightElement` is an element, so that gives me the confidence to remove this defensive bit 🙂